rustfmt-core/rustfmt-bin/src/bin/main.rs

```diff
@@ -35,25 +39,247 @@ fn main() {
+    ///
+    /// ```
+    /// rustfmt --file-lines '[
+    ///    {{\"file\":\"src/lib.rs\",\"range\":[7,13]}},
+    ///    {{\"file\":\"src/lib.rs\",\"range\":[21,29]}},
+    ///    {{\"file\":\"src/foo.rs\",\"range\":[10,11]}},
+    ///    {{\"file\":\"src/foo.rs\",\"range\":[15,15]}}]'
+    /// ```
+    ///
+    /// would format lines `7-13` and `21-29` of `src/lib.rs`, and lines `10-11`,
+    /// and `15` of `src/foo.rs`. No other files would be formatted, even if they
+    /// are included as out of line modules from `src/lib.rs`.
+    #[cfg_attr(nightly, structopt(long = "file-lines", default_value = "null"))]
+    #[cfg_attr(not(nightly), structopt(skip))]
+    file_lines: FileLines,
+
+    /// Error if unable to get comments or string literals within max_width,
+    /// or they are left with trailing whitespaces (unstable).
+    #[cfg_attr(nightly, structopt(long = "error_on_unformatted"))]
+    #[cfg_attr(not(nightly), structopt(skip))]
+    error_on_unformatted: bool,
+
+    // Positional arguments.
+
+    #[structopt(parse(from_os_str))]
+    files: Vec<PathBuf>,
+}
+
+#[derive(Debug, Clone)]
+struct InlineConfig(HashMap<String, String>, bool /* is help */);
+
+impl InlineConfig {
+    fn is_help(&self) -> bool {
+        self.1
+    }
+}
+
+impl FromStr for InlineConfig {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.trim() == "help" {
+            return Ok(InlineConfig(HashMap::default(), true));
+        }
+
+        s.split(',')
+            .map(
+            |key_val| match key_val.char_indices().find(|(_, ch)| *ch == '=') {
+                Some((middle, _)) => {
+                    let (key, val) = (&key_val[..middle], &key_val[middle + 1..]);
+                    if !Config::is_valid_key_val(key, val) {
+                        Err(format_err!("invalid key=val pair: `{}`", key_val))
+                    } else {
+                        Ok((key.to_string(), val.to_string()))
+                    }
+                }
+
+                None => Err(format_err!(
+                        "--config expects comma-separated list of key=val pairs, found `{}`",
+                        key_val
+                    )),
+            },
+        ).collect::<Result<HashMap<_, _>, _>>()
+            .map(|map| InlineConfig(map, false))
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum PrintConfig {
+    Default,
+    Minimal,
+    Current,
+}
+
+impl FromStr for PrintConfig {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "default" => Ok(PrintConfig::Default),
+            "minimal" => Ok(PrintConfig::Minimal),
+            "current" => Ok(PrintConfig::Current),
+            _ => Err(format!("expected one of [current,default,minimal], found `{}`", s)),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum Emit {
+    Files,
+    Stdout,
+    Checkstyle,
+    Json,
+}
+
```

Comment:
Is the removal of `Coverage` emit mode intentional?

Reply:
It was not, though I am planning to remove the coverage mode in the further PRs, so I might as well leave this as is.

Reply:
Yeah that makes sense